Minor Vector cleanup

l0rinc opened the pull request:
My original intention didn't work out, but since I created these simple style unification commits, I though I'll share them. :)

l0rinc committed:
Removed redundant casts from Vector

l0rinc committed:
Unified masks in Vector

l0rinc committed:
Deleted leftover code-comments from Vector

Ichoran commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -88,7 +83,7 @@ override def companion: GenericCompanion[Vector] = Vector
   override def lengthCompare(len: Int): Int = length - len
 
-  private[collection] final def initIterator[B >: A](s: VectorIterator[B]) {
+  private[collection] def initIterator[B >: A](s: VectorIterator[B]) {
```
If we can get away with finality, we should leave it in.  In general, immutable data structures do not lend themselves well to subclassing.

l0rinc replied:
the whole class is [final](https://github.com/scala/scala/pull/5545/files/228da21af702a51e8dc59956fcf61a0de3957726#diff-59f3462485b74027de4fd5e9febcc81bR64), does it make sense to keep final methods also?

Ichoran replied:
Ah.  No.  Sorry, I missed that.  I had somehow assumed, without any evidence, that it was in VectorPointer.

Ichoran commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -100,10 +95,7 @@ override def companion: GenericCompanion[Vector] = Vector
-  // can still be improved
-  override /*SeqLike*/
-  def reverseIterator: Iterator[A] = new AbstractIterator[A] {
+  override /*SeqLike*/ def reverseIterator: Iterator[A] = new AbstractIterator[A] {
```
This is harder to read.  It's better the original way.

l0rinc replied:
in other places it was formatted like this. should I change the rest also, or just revert this?

Ichoran replied:
If you don't mind changing the others, may as well change them.  The `override /* SeqLike */` thing is arguably the least important part of the signature, so having it push other stuff out of view on the right, potentially, is a bad idea.  Two lines solve the problem nicely.

Ichoran commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -286,61 +271,46 @@ override def companion: GenericCompanion[Vector] = Vector
-        val shiftBlocks = freeSpace >>> 5*(depth-1) // number of top-level blocks
+        val freeSpace = ((1 << (5 * depth)) - endIndex) // free space at the right given the current tree-structure depth
+        val shift = (freeSpace & ~((1 << (5 * (depth - 1))) - 1)) // number of elements by which we'll shift right (only move at top level)
+        val shiftBlocks = (freeSpace >>> (5 * (depth - 1))) // number of top-level blocks
```
Spaces are good.  Gratuitious parens around the entire expression are not good--just visual clutter that doesn't aid anything.  Internal parens that make one not rely upon knowing operator precedence are good.

l0rinc replied:
understood, will remove external ones in other places also :)

Ichoran commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -368,56 +336,41 @@ override def companion: GenericCompanion[Vector] = Vector
-        val shiftBlocks = startIndex >>> 5*(depth-1)
-
-        //println("----- appendBack " + value + " at " + endIndex + " reached block end")
+        val shift = startIndex & ~((1 << 5 * (depth - 1))-1)
```
Missed spaces around the last `-`

Ichoran commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -368,56 +336,41 @@ override def companion: GenericCompanion[Vector] = Vector
-
-        //println("----- appendBack " + value + " at " + endIndex + " reached block end")
+        val shift = startIndex & ~((1 << 5 * (depth - 1))-1)
+        val shiftBlocks = startIndex >>> 5 * (depth - 1)
```
Should surround multiplication with parens for clarity

Ichoran commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -609,20 +555,7 @@ override def companion: GenericCompanion[Vector] = Vector
-    s.cleanLeftEdge(cutIndex-shift)
-    s
-*/
+    val shift = (cutIndex & ~((1 << (5 * d)) - 1))
```
Don't need outer parens

Ichoran commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -639,13 +572,8 @@ override def companion: GenericCompanion[Vector] = Vector
     val xor = startIndex ^ (cutIndex - 1)
     val d = requiredDepth(xor)
-    val shift = (startIndex & ~((1 << (5*d))-1))
+    val shift = (startIndex & ~((1 << (5 * d)) - 1))
```
Don't need outer parens

Ichoran commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -832,7 +757,7 @@ private[immutable] trait VectorPointer[T] {
       } else
       if (xor < (1 << 10)) { // level = 1
-        display0 = display1((index >> 5) & 31).asInstanceOf[Array[AnyRef]]
+        display0 = display1((index >>  5) & 31).asInstanceOf[Array[AnyRef]]
```
Added an extra space by mistake?

Ichoran commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -1123,16 +1041,13 @@ private[immutable] trait VectorPointer[T] {
     // requires structure is clean and at pos oldIndex,
     // ensures structure is dirty and at pos newIndex and writable at level 0
     private[immutable] final def gotoFreshPosWritable0(oldIndex: Int, newIndex: Int, xor: Int): Unit = { // goto block start pos
       if (xor < (1 << 5)) { // level = 0
```
Should leave a comment that we're already at the block start.

Ichoran commented:
This looks like a welcome improvement overall; just a few aspects that could be even tidier if you're tidying anyway.

l0rinc committed:
Applied suggestions to Vector cleanup

paplorinc commented:
Thanks @Ichoran!
I've applied your suggestions and rebased the commit, hoping the build will pass with this base.
I've also changed all signed left shifts to unsigned ones, since all indices are inherently positive.

l0rinc commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -888,35 +888,35 @@ private[immutable] trait VectorPointer[T] {
-        display3((index >> 15) & 31) = display2
-        display2((index >> 10) & 31) = display1
-        display1((index >>  5) & 31) = display0
+        display5((index >>> 25) & 31) = display4
```
indices are inherently positive or zero

Ichoran replied:
If there is error-checking to make sure they're positive or zero, then there is no reason to bother changing. If there is not error-checking, an exception could be replaced with a silently wrong result.  I think they should stay signed!

l0rinc replied:
The error checking is in [checkRangeConvert](https://github.com/paplorinc/scala/blob/4279bbc9f467ef6d8a0ddf4190f2dd947f13d580/src/library/scala/collection/immutable/Vector.scala#L120) and `>>>` was already used in many other places, it's the reason why I unified it: https://github.com/paplorinc/scala/blob/4279bbc9f467ef6d8a0ddf4190f2dd947f13d580/src/library/scala/collection/immutable/Vector.scala#L482

l0rinc commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -414,22 +420,30 @@ override def companion: GenericCompanion[Vector] = Vector
-    val a2 = new Array[AnyRef](array.length)
-    java.lang.System.arraycopy(array, 0, a2, 0, right)
-    a2
+  private[immutable] def copyLeft(array: Array[AnyRef], right: Int): Array[AnyRef] = {
```
to unify with `copyOf`

paplorinc commented:
@szeiger, not sure how to fix this unrelated [MiMa](https://scala-ci.typesafe.com/job/scala-2.12.x-validate-test/3561/console) failure, what should I rebase againts?

l0rinc commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -113,22 +106,18 @@ override def companion: GenericCompanion[Vector] = Vector
   private def checkRangeConvert(index: Int) = {
     val idx = index + startIndex
-    if (0 <= index && idx < endIndex)
+    if (index >= 0 && idx < endIndex)
```
@Ichoran, this makes sure all indices are strictly non-negative.
(switched the comparison order to make it more intuitive - it's also easier to see that different variables are compared for boundaries)

szeiger commented:
There's nothing to rebase against yet, all 2.12.x builds are currently broken. I've created https://github.com/scala/scala/pull/5549 to fix this.

paplorinc commented:
@szeiger, I still have a MiMa failure (not sure where to view the failure's reason).
Is it because of the deleted [`final` method modifier in the `final` class](https://github.com/scala/scala/pull/5545#pullrequestreview-9292947)?

szeiger commented:
@paplorinc Go to the details of `validate-test`, select `Console Output` and `Full Log`, search for "binary compatibility":

```
/home/jenkins/workspace/scala-2.12.x-validate-test/build/pack/lib/scala-library.jar
Found 4 binary incompatibilities (1 were filtered out)
======================================================
 * method debug()Unit in class scala.collection.immutable.VectorBuilder does
   not have a correspondent in current version
 * method debug()Unit in class scala.collection.immutable.VectorIterator
   does not have a correspondent in current version
 * method debug()Unit in class scala.collection.immutable.Vector does not
   have a correspondent in current version
 * method debug()Unit in interface scala.collection.immutable.VectorPointer
   does not have a correspondent in current version
```

Unlike most projects we require both backwards and forwards binary compatibility for Scala itself, so you can never add or remove a method.

Some cases are really not binary compatible but we're OK with it because it only affects internal code that could not be accessed from outside the standard library without subverting Scala's access restrictions. Package-private elements generally fall into this category. Once you have convinced yourself that the binary-incompatible changes are harmless you can whitelist them. See the log for the ready-to-copy-and-paste whitelisting configuration. Note that it's always shown under backwards compatibility even when it really affects forwards compatibility (and therefore has to go into `bincompat-forward-whitelist.conf`).

l0rinc committed:
Deleted leftover debug method from Vector

l0rinc committed:
Changed >> to >>> in Vector to unify style

Unified, since indices are always positive and unsigned shift was already used in other places

paplorinc commented:
Thank you @szeiger, had to remove some things, but now it's finally green :/.
@Ichoran, please review :)

Ichoran commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -351,73 +327,56 @@ override def companion: GenericCompanion[Vector] = Vector
-        val shiftBlocks = startIndex >>> 5*(depth-1)
-
-        //println("----- appendBack " + value + " at " + endIndex + " reached block end")
+        val shift       = startIndex & ~((1 << (5 * (depth - 1))) - 1)
```
This is not sufficiently similar to the next computation for alignment to be a good idea.  It suggests that the two are closely related, when in fact they are in detail rather different computations.  So I would just go `shift = startIndex` without the extra spaces.

Ichoran commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -639,29 +575,22 @@ override def companion: GenericCompanion[Vector] = Vector
-   with VectorPointer[A @uncheckedVariance] {
+  extends AbstractIterator[A]
+    with Iterator[A]
+    with VectorPointer[A @uncheckedVariance] {
```
This style makes it hard to tell at a glance where the class signature ends and where the body begins.  The style is all over the place in practice, but I recommend left-justifying everything; vertical whitespace should always be used to separate classes so there is no need to indent them to indicate that they all belong to `VectorIterator`.  Failing that, the previous version was preferable because at least all the extended classes/traits are lined up.

Ichoran commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -799,98 +725,81 @@ private[immutable] trait VectorPointer[T] {
+      } else if (xor < (1 << 25)) { // level = 4
+        display4((index >>> 20) & 31).asInstanceOf[Array[AnyRef]]((index >>> 15) & 31).asInstanceOf[Array[AnyRef]]((index >>> 10) & 31).asInstanceOf[Array[AnyRef]]((index >>>  5) & 31).asInstanceOf[Array[AnyRef]](index & 31).asInstanceOf[T]
+      } else if (xor < (1 << 30)) { // level = 5
+        display5((index >>> 25) & 31).asInstanceOf[Array[AnyRef]]((index >>> 20) & 31).asInstanceOf[Array[AnyRef]]((index >>> 15) & 31).asInstanceOf[Array[AnyRef]]((index >>> 10) & 31).asInstanceOf[Array[AnyRef]]((index >>> 5) & 31).asInstanceOf[Array[AnyRef]](index & 31).asInstanceOf[T]
```
This is an awful mess even with the improvements unless you have a very wide screen.  If you're up for it maybe you can rewrite it in the style
```
display5((index >>> 25) & 31).asInstanceOf[Array[AnyRef]].
   apply((index >>> 20) & 31).asInstanceOf[Array[AnyRef]].
   apply((index >>> 15) & 31) . . .
```
The alternative--lining up the matching part of the displays--pushes the earlier ones so far right that they disappear, so that's probably not a good idea.

l0rinc replied:
how about
```scala
var array = display5 
array = array((index >>> 25) & 31).asInstanceOf[Array[AnyRef]]
array = array((index >>> 20) & 31).asInstanceOf[Array[AnyRef]]
array = array((index >>> 15) & 31).asInstanceOf[Array[AnyRef]]
array = array((index >>> 10) & 31).asInstanceOf[Array[AnyRef]]
array = array((index >>>  5) & 31).asInstanceOf[Array[AnyRef]]
array(index & 31).asInstanceOf[T]
```
or would that be slower than the direct reference?

Ichoran commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -1118,117 +1023,83 @@ private[immutable] trait VectorPointer[T] {
-
-}
-
+}
```
Add a newline at the end to keep things happy that like newlines at the end!

l0rinc replied:
:)) sir, yes sir!

l0rinc committed:
Applied further cleanup to Vector

Ichoran commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -799,98 +727,102 @@ private[immutable] trait VectorPointer[T] {
+        array = array((index >>> 10) & 31).asInstanceOf[Array[AnyRef]]
+        array = array((index >>>  5) & 31).asInstanceOf[Array[AnyRef]]
+        array(index & 31).asInstanceOf[T]
+      } else {                      // level = 6
```
I have witnessed cases where allocating a temporary var produces worse performance than chaining methods, and at this point in time I think chaining is slightly more idiomatic than repeated assignment to vars.  So I'd just chain them as I indicated before.  If you don't like the `apply`, you can do it this way:
```
array((index >>> 25) & 31).
  asInstanceOf[Array[AnyRef]]((index >>> 20) & 31).
  asInstanceOf[Array[AnyRef]]((index >>> 15) & 31).
  ...
```
You miss the alignment on the first shift, but I guess you could add whitespace until it was aligned.

l0rinc replied:
Done, resolved it via parenthesizing everything to make it multilinable, making alignment trivial also :)

Ichoran reviewed (approved):
LGTM

paplorinc commented:
@retronym, could you please review this also, it's related to https://issues.scala-lang.org/browse/SI-10049 as well?

paplorinc commented:
@Ichoran, @retronym, rebased to fix conflict in whitelist.

szeiger commented:
/nothingtoseehere

The CLA checker had some hickups recently. CLA confirmed.

SethTisue commented:
thanks @paplorinc! you're killing it lately

paplorinc commented:
Thank you @Ichoran, @szeiger and @SethTisue! :)
